### PR TITLE
feat(hash): compute & verify file checksums

### PR DESCRIPTION
**Pull request by genesiscz**
## What

`tools hash` — a cross-platform CLI to compute and verify file checksums, backed by the already-installed `hash-wasm` dependency. One command replaces the per-platform mix of `shasum`, `md5sum`, and `b3sum` (blake3 isn't available by default on macOS).

## CLI surface

```
tools hash [options] <files...>          # compute mode (default)
tools hash --check <file> [options]      # verify mode
```

- `-a, --algo <md5|sha1|sha256|sha512|blake3>` — algorithm (default `sha256`, validated via commander `.choices`).
- `-c, --check <file>` — verify a checksum file instead of computing (mutually exclusive with positional files).
- `--quiet` — in `--check` mode, print only `FAILED` lines.

`<files...>` are paths or glob patterns (quote globs; expanded via the `glob` dep with `nodir: true`, de-duped, deterministic order).

## Behavior

- **Streaming**: files are read via `Bun.file(path).stream()` and fed to `hash-wasm`'s `IHasher` chunk-by-chunk (`init` → `update` → `digest`). The whole file is never buffered, so multi-GB files work.
- **Coreutils-compatible output**: `<lowercase-hex><two spaces><path>` — byte-identical to `shasum`/`md5sum`/`b3sum`, so output feeds straight back into `--check` (and vice-versa). Verified against system `shasum -a 256`.
- **`--check`**: tolerant parser (skips blank/`#` lines, accepts 1+ spaces and the GNU `*` binary marker, lowercases hex), recomputes each entry, prints `<path>: OK` / `<path>: FAILED` / `<path>: FAILED (could not read)`, with a summary to stderr.
- **Exit codes**: 0 only when every operation succeeds; 1 on any mismatch, unreadable file, or zero glob matches.
- **Output discipline**: results via `out.println` (stdout only); status/errors/summaries via `out.error` / `out.log.*` / `logger` (stderr). `await out.flush()` precedes every `process.exit()` so redirected output (`> sums.sha256`) is never truncated mid-drain.

## Structure

- Pure, testable lib in `src/hash/lib/` — `algorithms.ts` (registry + `createHasher`), `hash-stream.ts` (`hashChunks`/`hashBuffer`), `checksum-file.ts` (format/parse/summarize). No commander, `out`, `process.exit`, or clock reads.
- Thin commander entrypoint in `src/hash/index.ts`, ending in `await runTool(program, { tool: "hash" })`.
- Hermetic `bun:test` in `src/hash/hash.test.ts` — known-vector digests per algo, chunked==one-shot equality, async-iterable feed, tolerant-parser cases, and an end-to-end verify over `os.tmpdir()` files (no network, no clipboard, no live-repo state).

## Verification (local)

- `bunx biome check src/hash` — exit 0.
- `bun test src/hash` — 18 pass / 0 fail.
- `./tools hash --help` — exit 0.
- Compute matches `shasum -a 256`; redirect round-trip + verify/tamper/quiet/glob/no-match all behave per spec.

No new npm dependencies; no `console.log`; no `JSON.*`; no `any`; no hardcoded user paths.

## Summary by CodeRabbit

* **New Features**
  * Added a new `hash` command for computing and verifying file checksums
  * Supports multiple hashing algorithms (MD5, SHA1, SHA256, SHA512, BLAKE3)
  * Verification mode to validate checksums against a checksum file
  * Glob pattern support for processing multiple files
  * Quiet mode to show only verification failures

* **Tests**
  * Added comprehensive test suite for hashing and verification functionality

### CODE REVIEW

**Comment by genesiscz** [comment]
@coderabbitai review

**Comment by genesiscz** [comment]
@coderabbitai review

**Comment by genesiscz** [comment]
@coderabbitai review

**Comment by genesiscz** [comment]
@coderabbitai review

**Comment by genesiscz** [comment]
@coderabbitai review

File: src/hash/index.ts [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+        for (const match of matches) {
+            if (!seen.has(match)) {
+                seen.add(match);
+                result.push(match);
+            }
+        }
+    }
+
+    return result;
+}
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/hash/index.ts:61]: @gemini-code-assist Fixed in [635bd01](https://github.com/genesiscz/GenesisTools/commit/635bd012038e9ea769cbb8afda33a8bc87703e6d) - expandGlobs now passes literal (non-glob) paths through unchanged, so a missing literal file flows into the read loop and surfaces as 'could not read' with a non-zero exit instead of being silently dropped. One tweak to the suggestion: the glob-detection class is [*?[]{}] (no whitespace), since treating spaces as glob chars would misclassify paths that contain spaces.

File: src/hash/lib/checksum-file.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        const match = line.match(/^([0-9a-fA-F]+)\s+\*?(.+)$/);
+        if (!match) {
+            continue;
+        }
+
+        entries.push({ hex: match[1].toLowerCase(), path: match[2] });
+    }
+
+    return entries;
+}
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/hash/lib/checksum-file.ts:39]: @gemini-code-assist Won't fix - both points don't hold against the current code. (1) The greedy whitespace match is correct and the suggestion regresses it: it consumes all separator whitespace so no leading space leaks into the path. For a line with hash deadbeef, a 3-space separator, then 'my file.txt', the current regex yields path 'my file.txt', but the suggested separator (at most two spaces) yields ' my file.txt' with a leaked leading space. The existing tests 'preserves spaces inside the path' and 'tolerates a single space and the GNU binary marker' already lock this in. (2) Verbatim paths are intentional and coreutils-compatible: GNU sha256sum -c does not normalize backslash to forward slash, and backslash is a legal Unix filename character, so unconditional replacement would corrupt valid Unix paths. Leaving the parser as-is.

File: src/hash/index.ts [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+        out.error(`hash: cannot open checksum file: ${checkFile}`);
+        logger.error({ checkFile, error }, "hash: failed to read checksum file");
+        return 1;
+    }
+
+    const entries = parseChecksumFile(text);
+    if (entries.length === 0) {
+        out.error(`hash: no checksum entries found in ${checkFile}`);
+        return 1;
+    }
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/hash/index.ts:117]: @gemini-code-assist Fixed in [635bd01](https://github.com/genesiscz/GenesisTools/commit/635bd012038e9ea769cbb8afda33a8bc87703e6d) - runCheck now performs an O(1)-per-entry length pre-check before verifyEntries (md5=32, sha1=40, sha256/blake3=64, sha512=128) and fast-fails with a clear message when the parsed hash lengths don't match the selected algorithm, avoiding rehashing every file. Note sha256 and blake3 are both 64 chars, so the guard only catches gross algorithm mismatches, which is the intent.

**Comment by genesiscz** [comment]
@coderabbitai review

**Comment by genesiscz** [comment]
@coderabbitai review

**Comment by genesiscz** [comment]
@coderabbitai review

File: src/hash/hash.test.ts [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+async function verifyEntries(dir: string, algo: HashAlgo, entries: ChecksumEntry[]): Promise<VerifyResult[]> {
+    const results: VerifyResult[] = [];
+    for (const entry of entries) {
+        try {
+            const bytes = new Uint8Array(await Bun.file(join(dir, entry.path)).arrayBuffer());
+            const actual = await hashBuffer(algo, bytes);
+            results.push({ path: entry.path, ok: actual === entry.hex });
+        } catch {
+            results.push({ path: entry.path, ok: false, unreadable: true });
+        }
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/hash/hash.test.ts:140]: @coderabbitai By design - this is a test helper whose catch is the deliberately-asserted 'unreadable file' branch: the e2e test seeds a missing c.txt and asserts cResult.unreadable === true, so the swallow IS the behavior under test, not a masked regression (a real throw while hashing a valid file would still fail the OK/FAILED assertions). Bare catch {} in test fixtures is the established repo convention (indexer, search, integration, paths, io-edge-cases tests all do the same). The no-bare-catch guideline targets production diagnostics via @app/logger, which test files do not wire up.

File: src/hash/index.ts [hunk 5]
```diff
@@ -0,0 +1,195 @@
+import { logger, out } from "@app/logger";
+import { isInteractive, runTool, suggestCommand } from "@app/utils/cli";
+import { Command, Option } from "commander";
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/hash/index.ts:3]: @coderabbitai Won't fix - the interactive tool-selector convention applies to the ROOT tools executable (which lists all tools when run bare), not to individual src/tool/index.ts entrypoints. Per-tool entrypoints validate their own args and error with a suggestCommand hint, which is exactly what hash does. Sibling single-purpose tools follow the identical pattern: say/index.ts and transcribe/index.ts both print an error plus suggestCommand and exit(1) on missing args rather than launching a selector. hash also already gates the hint on isInteractive(). No per-tool selector exists or is required.

File: src/hash/index.ts [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+        logger.warn({ patterns }, "hash: zero files matched");
+        return 1;
+    }
+
+    let failures = 0;
+    for (const file of files) {
+        try {
+            const hex = await hashFile(algo, file);
+            out.println(formatChecksumLine(hex, file));
+        } catch (error) {
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/hash/index.ts:77]: @coderabbitai Won't fix - out.print and out.println ARE sanctioned stdout writers, not just out.result. The Out interface documents println as the 'console.log-ergonomic stdout path' (it routes through emitResult to writeStdout) and print as the raw stdout path; out.result is specifically for serializing ONE machine result via asResult, which is wrong for a per-file streaming loop. The CI logging-guard only forbids logger.* being used as the result channel, not non-result stdout writers, and out.println is used this way in 205 places across the repo. The success summary at line 146 correctly uses out.log.success (stderr) since it's a status banner, not the result. stdout discipline is already honored.
